docs/data-source.md

```diff
@@ -0,0 +1,131 @@
+Retcon Data Sources
+===================
+
+This document describes the interface to be implemented by a Retcon data source
+program.
+
+Overview
+--------
+
+A *data source* is a program which Retcon can invoke when it needs to interact
+with an external system. Each invocation of a data source processes a single
+operation and they can be implemented in any language.
+
+Data sources must implement the four CRUD operations:
+
+- Create a new document;
+- Read an existing document;
+- Update an existing document; and
+- Delete an existing document.
+
+Formats
+-------
+
+The following data will be communicated between Retcon and a data source program:
+
+- JSON documents describing data items.
+
+- Text strings representing foreign keys.
+
+JSON documents must follow the JSON specification and not include third-party
+extensions (like `Date` values).
+
+Foreign keys should be non-empty sequences of the following ASCII-safe
+characters only:
+
+- uppercase alphabetic;
+- lowercase alphabetic ;
+- numeric;
+- `.` `-` `,` `_`;
+- space.
+
+Interface
+---------
+
+Data source programs should implement the following interface.
+
+### General
+
+All operations must exit with status 0 when they complete successfully. They
+should produce any output required to the stdout stream and may produce
+debugging messages to stderr. These messages will be logged by Retcon at the
+DEBUG log level.
+
+All operations must exit with status 1 when they failed because they could not
+communicate with the external system. They must print a useful error message to
+their stderr stream. These messages will be logged by Retcon at the EMERG log
+level.
```

Comment:
Why EMERG? retcond won't immediately cease to function if a datasource exits with error, which is normally what an EMERG message indicates (e.g., kernel panic). Maybe ALERT if such an error is always something requiring administrator intervention, otherwise just ERROR?